polars/polars-core/src/chunked_array/logical/struct_/mod.rs

```diff
@@ -173,58 +177,46 @@ impl StructChunked {
             field,
             chunks: vec![arrow_array],
             null_count: 0,
+            total_null_count: 0,
         };
         out.set_null_count();
         out
     }
 
     fn set_null_count(&mut self) {
-        let mut null_count = 0;
+        // Count both the total number of nulls and the rows where everything is null
+        let (mut null_count, mut total_null_count) = (0, 0);
         let chunks_lens = self.fields()[0].chunks().len();
 
-        // fast path
-        // we early return if a column doesn't have nulls
-        for i in 0..chunks_lens {
-            for s in self.fields() {
-                let arr = &s.chunks()[i];
-                let has_nulls = arr.null_count() > 0 || matches!(s.dtype(), DataType::Null);
-                if !has_nulls {
-                    self.null_count = 0;
-                    return;
-                }
-            }
-        }
-
-        // slow path
-        // we bitand every null validity bitmask to determine
-        // in which rows all values are null
+        // A row is null if all values in it are null, so we bitor every validity bitmask since a
+        // single valid entry makes that row not null. We can also save some work by not bothering
+        // to bitor fields that would have all 0 validities (Null dtype or everything null). Note
+        // that since we keep track of the total null count as well, we can't break early, but we
+        // are only dealing with the validity masks when we absolutely have to.
         for i in 0..chunks_lens {
-            let mut validity_agg = None;
-
-            let mut all_null_array = true;
+            let mut validity_agg: Option<arrow::bitmap::Bitmap> = None;
+            let mut n_nulls = None;
             for s in self.fields() {
                 let arr = &s.chunks()[i];
-
-                if !matches!(s.dtype(), DataType::Null) {
-                    all_null_array = false;
-                    match (&validity_agg, arr.validity()) {
-                        (Some(agg), Some(validity)) => validity_agg = Some(validity.bitand(agg)),
-                        (None, Some(validity)) => validity_agg = Some(validity.clone()),
-                        _ => {}
+                let nc = arr.null_count();
+                match (n_nulls, arr.validity(), nc == arr.len() && nc > 0) {
+                    (_, _, true) => total_null_count += arr.len(),
+                    (Some(0), _, _) => {}
+                    (_, Some(v), _) => {
+                        validity_agg =
+                            validity_agg.map_or(Some(v.clone()), |agg| Some(v.bitor(&agg)));
+                        n_nulls = Some(validity_agg.as_ref().unwrap().unset_bits());
                     }
+                    (_, None, _) => n_nulls = Some(0),
```

Comment:
We should break early here.

Reply:
Can't with the current algorithm, see above. But if you're ok with potentially two full passes over all the fields where the first one is very cheap, we can. After writing the above explanation it does sound like that will be better.

Reply:
Now that we count total nulls before, it will break.